src/error/src/anyhow.rs

```diff
@@ -154,4 +154,30 @@ macro_rules! def_anyhow_newtype {
         )*
     };
 }
-pub use def_anyhow_newtype;
+
+/// Define a newtype + it's variant in the specified type.
+/// This is useful when you want to define a new error type,
+/// but also want to define a variant for it in another enum.
+#[macro_export]
+macro_rules! def_anyhow_variant {
+    (
+        $(#[$attr:meta])* $vis:vis $name:ident,
+        $enum_name:ident $variant_name:ident
+        $(, $from:ty => $context:tt)* $(,)?
+    ) => {
+        def_anyhow_newtype! {
+            $(#[$attr])* $vis $name
+            $(, $from => $context)*
+        }
+
+        $(
+            impl From<$from> for $enum_name {
+                fn from(error: $from) -> Self {
+                    todo!()
+                }
+            }
+        )*
```

Comment:
I'm uncertain if this is necessary or a best practice. I believe we can have those functions or streams interact with external libraries to directly return `ExternalSystemError` as their error type, so that we only need to call `map_err(Into::into)` once to convert the error type into `BatchError`.

Reply:
If we directly embedded external system errors directly into `BatchError`, as was done before this PR, we would have `From` instances implemented for these errors.

```rust
pub enum BatchError {
    ...
    #[error(transparent)]
    Iceberg(
        #[from]
        #[backtrace]
        iceberg::Error,
    ),

    #[error(transparent)]
    Parquet(
        #[from]
        #[backtrace]
        ParquetError,
    ),

    #[error(transparent)]
    Postgres(
        #[from]
        #[backtrace]
        tokio_postgres::Error,
    ),
    ...
}
```

So I don't see an issue with deriving it as well, when using a nested error type.
It's also more ergonomic and readable. `map_err(Into::into)` looks strange. If I saw `map_err(Into::into)` I would wonder we do not directly coerce the error type using `?`.

Reply:
You're right and I found myself wrong in what I said.

My only concern is that, there can be conflicts if there's an error type with different semantics between when it's directly embedded in `BatchError` and when it's a variant of `ExternalSystemError`. However, it seems not possible in this case, as long as we're only including concrete types from external crates in the wrapper type, rather than more general types that require additional context, like `FromIntError`.

Reply:
I think for extenral system errors which could be embedded inside both `ExternalSystemError` and `BatchError`, we can instead not embed them inside `BatchError`, but use `.context` to specify more info around them.